Rework MethodInheritance.fixMethodInheritance to "soft" dependencies

### TestNG Version

7.0.0, 7.3.0

### Expected behavior

I expect `@BeforeSuite(groups=...)` should add configuration methods to a "configuration group", and I expect multiple different classes should be able to "add a configuration method" to a group.

My use case is a database-related test: "prepare configuration", "upload configuration", "execute test". The upload step might take 1-2 minutes, so I batch the upload: I gather the bits from all the test classes in the suite, and I execute "upload configuration to db" just once.

The structure of the tests is as follows:

```java
import org.testng.annotations.BeforeSuite;

public class Base {
    @BeforeSuite(groups = "prepareConfig")
    public void prepareConfig() {
        // This method prepares the baseline configuration
        // Other classes might contribute to it via their own @BeforeSuite(groups = "prepareConfig")
        System.out.println("prepareConfig");
    }

    @BeforeSuite(groups = "uploadConfigToDatabase", dependsOnGroups = "prepareConfig")
    public void uploadConfigToDatabase() {
        // This method uploads the configuration to database. In most of the cases, it is the only method in uploadConfigToDatabase group
        System.out.println("uploadConfigToDatabase");
    }

    @BeforeSuite(groups = "verifyConfigurationAfterInstall", dependsOnGroups = "uploadConfigToDatabase")
    public void verifyConfigurationAfterInstall() {
        // Verify the database before tests start executing
        // Other classes can contribute their own checks via @BeforeSuite(groups = "verifyConfigurationAfterInstall")
        System.out.println("verifyConfigurationAfterInstall");
    }
}
```

```java
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Test;

public class Test1 extends Base {
    @BeforeSuite(groups = "prepareConfig")
    public void prepareConfigForTest1() {
        System.out.println("prepareConfigForTest1");
    }

    @Test
    public void test1() {
    }
}
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<suite name="Groups test" parallel="false">
  <test name="Test groups">
    <classes>
      <class name="Test1">
        <methods><include name="test1"/></methods>
      </class>
    </classes>
  </test>
</suite>
```

I expect that `Test1#prepareConfigForTest1` participates in `prepareConfig` configuration group, so I expect the following execution order:

```
Base#prepareConfig
Test1#prepareConfigForTest1
Base#uploadConfigToDatabase
Base#verifyConfigurationAfterInstall

Test1#test1
```

In other words, I expect that the full set of dependencies is as follows:

```
Base#prepareConfig: no dependencies

Test1#prepareConfigForTest1
  "soft" depends on all BeforeSuite in Base (because Test1 inherits Base class):
      Base#prepareConfig, Base#uploadConfigToDatabase, Base#verifyConfigurationAfterInstall
  "hard" depends on: empty

Base#uploadConfigToDatabase
  "hard" depends on: all configuration methods in prepareConfig group: Base#prepareConfig, Test1#prepareConfigForTest1

Base#verifyConfigurationAfterInstall
  "hard" depends on: all configuration methods in uploadConfigToDatabase: Base#uploadConfigToDatabase
```

Then TestNG should make the topological sort taking all hard constraints into consideration, and it should fulfill soft constraints on the best effort basis.

### Actual behavior

TestNG adds "implicit" dependency (via `org.testng.internal.MethodInheritance#fixMethodInheritance`)  of "`Test1#prepareConfigForTest1` depends on `Base#verifyConfigurationAfterInstall`" which results in a configuration cycle, so the engine fails as follows

```
Caused by: org.testng.TestNGException: 
The following methods have cyclic dependencies:
Base.verifyConfigurationAfterInstall()[pri:0, instance:null]
Test1.prepareConfigForTest1()[pri:0, instance:null]
Base.uploadConfigToDatabase()[pri:0, instance:null]

	at org.testng.internal.Graph.topologicalSort(Graph.java:128)
	at org.testng.internal.MethodHelper.topologicalSort(MethodHelper.java:299)
	at org.testng.internal.MethodHelper.sortMethods(MethodHelper.java:382)
	at org.testng.internal.MethodHelper.collectAndOrderMethods(MethodHelper.java:71)
	at org.testng.internal.TestNGMethodFinder.findConfiguration(TestNGMethodFinder.java:200)
	at org.testng.internal.TestNGMethodFinder.getBeforeSuiteMethods(TestNGMethodFinder.java:79)
	at org.testng.TestClass.initMethods(TestClass.java:123)
	at org.testng.TestClass.init(TestClass.java:75)
	at org.testng.TestClass.<init>(TestClass.java:39)
	at org.testng.TestRunner.initMethods(TestRunner.java:457)
	at org.testng.TestRunner.init(TestRunner.java:336)
	at org.testng.TestRunner.init(TestRunner.java:289)
	at org.testng.TestRunner.<init>(TestRunner.java:180)
	at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:613)
	at org.testng.SuiteRunner.init(SuiteRunner.java:178)
	at org.testng.SuiteRunner.<init>(SuiteRunner.java:112)
	at org.testng.TestNG.createSuiteRunner(TestNG.java:1306)
	at org.testng.TestNG.createSuiteRunners(TestNG.java:1282)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1131)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.runTests(TestNGTestClassProcessor.java:141)
	at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.stop(TestNGTestClassProcessor.java:90)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
```


### Is the issue reproductible on runner?

The above exception is from Gradle 6.6, however, I'm quite sure it is reproducible from Shell runner as well.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2432: Rework MethodInheritance.fixMethodInheritance to "soft" dependencies (Krishnan Mahadevan)
 Fixed: GITHUB-2429: Seggregate Dependency Injection out as a clear implementation (Krishnan Mahadevan)
 Fixed: GITHUB-2435: getParameterIndex() always return 0 in test listener
 Fixed: GITHUB-2406: TestNG 7.3.0 transitive vulnerability CVE-2020-11022 and CVE-2020-11023 due to JQuery 3.4.1  (Krishnan Mahadevan)

--- a/src/main/java/org/testng/internal/MethodInheritance.java
+++ b/src/main/java/org/testng/internal/MethodInheritance.java
@@ -8,6 +8,7 @@ import java.util.Map;
 
 import java.util.Map.Entry;
 import java.util.Objects;
+import java.util.Optional;
 import org.testng.ITestNGMethod;
 import org.testng.collections.Lists;
 import org.testng.collections.Maps;
@@ -119,10 +120,8 @@ public class MethodInheritance {
     // Each bucket that has a list bigger than one element gets sorted
     //
     map.values().parallelStream()
+        .filter(l -> l.size() > 1)
         .forEach(l -> {
-          if (l.isEmpty() || l.size() == 1) {
-            return;
-          }
           // Sort them
           sortMethodsByInheritance(l, before);
 
@@ -136,6 +135,11 @@ public class MethodInheritance {
             ITestNGMethod m1 = l.get(i);
             for (int j = i + 1; j < l.size(); j++) {
               ITestNGMethod m2 = l.get(j);
+              String[] groups = Optional.ofNullable(m2.getGroups()).orElse(new String[] {});
+              if (groups.length != 0) {
+                //Do not resort to adding implicit depends-on if there are groups
+                continue;
+              }
               if (!equalsEffectiveClass(m1, m2) && !dependencyExists(m1, m2, methods)) {
                 Utils.log("MethodInheritance", 4, m2 + " DEPENDS ON " + m1);
                 m2.addMethodDependedUpon(MethodHelper.calculateMethodCanonicalName(m1));

--- a/src/test/java/test/configuration/issue2432/Base.java
+++ b/src/test/java/test/configuration/issue2432/Base.java
@@ -1,0 +1,17 @@
+package test.configuration.issue2432;
+
+import org.testng.annotations.BeforeSuite;
+
+public class Base {
+  @BeforeSuite(groups = "prepareConfig")
+  public void prepareConfig() {
+  }
+
+  @BeforeSuite(groups = "uploadConfigToDatabase", dependsOnGroups = "prepareConfig")
+  public void uploadConfigToDatabase() {
+  }
+
+  @BeforeSuite(groups = "verifyConfigurationAfterInstall", dependsOnGroups = "uploadConfigToDatabase")
+  public void verifyConfigurationAfterInstall() {
+  }
+}

--- a/src/test/java/test/configuration/issue2432/IssueTest.java
+++ b/src/test/java/test/configuration/issue2432/IssueTest.java
@@ -1,0 +1,38 @@
+package test.configuration.issue2432;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import static org.assertj.core.api.Assertions.assertThat;
+import org.testng.TestNG;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlClass;
+import org.testng.xml.XmlInclude;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlSuite.ParallelMode;
+import org.testng.xml.XmlTest;
+import test.InvokedMethodNameListener;
+import test.SimpleBaseTest;
+
+public class IssueTest extends SimpleBaseTest {
+
+  @Test(description = "GITHUB-2432")
+  public void ensureNoImplicitDependencyIsAddedWhenGroupsAreInvolved() {
+    XmlClass xmlClass = new XmlClass(Test1.class);
+    XmlInclude xmlInclude = new XmlInclude("test1");
+    xmlClass.setIncludedMethods(Collections.singletonList(xmlInclude));
+    XmlSuite xmlsuite = createXmlSuite("2432_suite");
+    xmlsuite.setParallel(ParallelMode.NONE);
+    XmlTest xmlTest = new XmlTest(xmlsuite);
+    xmlTest.setName("2432_test");
+    xmlTest.setXmlClasses(Collections.singletonList(xmlClass));
+    TestNG testng = create(xmlsuite);
+    InvokedMethodNameListener listener = new InvokedMethodNameListener();
+    testng.addListener(listener);
+    testng.run();
+    List<String> expected = Arrays.asList("prepareConfig", "prepareConfigForTest1", "uploadConfigToDatabase",
+        "verifyConfigurationAfterInstall", "test1");
+    assertThat(listener.getInvokedMethodNames()).containsExactlyElementsOf(expected);
+  }
+
+}

--- a/src/test/java/test/configuration/issue2432/Test1.java
+++ b/src/test/java/test/configuration/issue2432/Test1.java
@@ -1,0 +1,14 @@
+package test.configuration.issue2432;
+
+import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Test;
+
+public class Test1 extends Base {
+  @BeforeSuite(groups = "prepareConfig")
+  public void prepareConfigForTest1() {
+  }
+
+  @Test(groups="test")
+  public void test1() {
+  }
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -128,6 +128,7 @@
       <class name="test.testng195.AfterMethodTest" />
       <class name="test.regression.BeforeTestFailingTest"/>
       <class name="test.configuration.github1700.RunTest"/>
+      <class name="test.configuration.issue2432.IssueTest"/>
       <class name="test.aftergroups.AfterGroupsBehaviorTest"/>
       <class name="test.testng285.TestNG285Test" />
       <class name="test.failedreporter.FailedReporterTest" />

